### PR TITLE
ARROW-6807: [Java][FlightRPC] Expose gRPC service & client

### PR DESCRIPTION
**Pull request by rohit7195**
Creates a utility class that exposes the flight service & client so that multiple services
can be plugged into the same endpoint.
Changes access modifier of FlightClient ctor to package-private
Introduces unit test that binds multiple services to same server and allows multiple
clients to use the services.
Updates gRPC version to 1.24.0

### CODE REVIEW

File: java/flight-grpc/pom.xml [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <parent>
+    <artifactId>arrow-java-root</artifactId>
+    <groupId>org.apache.arrow</groupId>
+    <version>1.0.0-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>arrow-flight-grpc</artifactId>
+  <name>Arrow Flight GRPC</name>
```
Comment (emkornfield) [inline java/flight-grpc/pom.xml]: please add a description as well (see other packages).

File: java/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.util.concurrent.ExecutorService;
+
+import org.apache.arrow.flight.auth.ServerAuthHandler;
+import org.apache.arrow.memory.BufferAllocator;
+
+import io.grpc.ManagedChannel;
+
+/**
+ * Exposes Flight service & client.
```
Comment (emkornfield) [inline java/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java]:
```suggestion
 * Exposes Flight GRPC service & client.
```

File: java/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ */
+public class FlightGrpcUtils {
+
+  private FlightGrpcUtils() {}
+
+  public static FlightBindingService getFlightService(BufferAllocator allocator, FlightProducer producer,
```
Comment (emkornfield) [inline java/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java]: would createFlightService be more accurate?

Comment (emkornfield) [inline java/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java]: please add javadoc

File: java/flight-grpc/src/test/java/org/apache/arrow/flight/TestFlightGrpcUtils.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import io.grpc.inprocess.InProcessChannelBuilder;
+import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.stub.StreamObserver;
+
+/**
+ * Unit test which adds 2 services to same server end point.
+ */
+public class TestFlightGrpcUtils {
+
+  @Test(expected = RuntimeException.class)
```
Comment (emkornfield) [inline java/flight-grpc/src/test/java/org/apache/arrow/flight/TestFlightGrpcUtils.java]: use assertThrows instead of this.

File: java/flight-grpc/src/test/java/org/apache/arrow/flight/TestFlightGrpcUtils.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  public void testGrpcConnection() throws IOException {
```
Comment (emkornfield) [inline java/flight-grpc/src/test/java/org/apache/arrow/flight/TestFlightGrpcUtils.java]: please add more details on what exactly is being tested here.

File: java/flight/pom.xml [hunk 6]
```diff
@@ -95,6 +95,11 @@
       <artifactId>protobuf-java</artifactId>
       <version>3.5.1</version>
     </dependency>
+    <dependency>
```
Comment (emkornfield) [inline java/flight/pom.xml]: why is this necessary if we already import grpc?

**Comment by rohit7195** [comment]
@jacques-n @lidavidm @wesm

File: java/flight-grpc/src/test/java/org/apache/arrow/flight/TestFlightGrpcUtils.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    server.start();
+
+    final ManagedChannel managedChannel = InProcessChannelBuilder.forName(serverName)
+            .directExecutor()
+            .build();
+
+    final FlightClient flightClient = FlightGrpcUtils.getFlightClient(allocator, managedChannel);
+
+    final Iterable<ActionType> actionTypes = flightClient.listActions();
+    actionTypes.forEach(actionType -> System.out.println(actionType.toString()));
```
Comment (lidavidm) [inline java/flight-grpc/src/test/java/org/apache/arrow/flight/TestFlightGrpcUtils.java]: Is this the line that's intended to throw? Why is there more code after it? Also, shouldn't the exception thrown be of type `FlightRuntimeException`?

File: java/flight/pom.xml [hunk 8]
```diff
@@ -20,7 +20,7 @@
   <packaging>jar</packaging>
 
   <properties>
-    <dep.grpc.version>1.14.0</dep.grpc.version>
+    <dep.grpc.version>1.24.0</dep.grpc.version>
```
Comment (lidavidm) [inline java/flight/pom.xml]: This is a pretty big version jump...do we have any policy for bumping dependency versions?

File: java/flight-grpc/pom.xml (hunk 1 above)
Comment (rohit7195) [reply java/flight-grpc/pom.xml]: Done

File: java/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java (hunk 2 above)
Comment (rohit7195) [reply java/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java]: Done

File: java/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java (hunk 3 above)
Comment (rohit7195) [reply java/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java]: I agree.

Comment (rohit7195) [reply java/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java]: Done

File: java/flight-grpc/src/test/java/org/apache/arrow/flight/TestFlightGrpcUtils.java (hunk 4 above)
Comment (rohit7195) [reply java/flight-grpc/src/test/java/org/apache/arrow/flight/TestFlightGrpcUtils.java]: Done

File: java/flight-grpc/src/test/java/org/apache/arrow/flight/TestFlightGrpcUtils.java (hunk 5 above)
Comment (rohit7195) [reply java/flight-grpc/src/test/java/org/apache/arrow/flight/TestFlightGrpcUtils.java]: Done

File: java/flight/pom.xml (hunk 6 above)
Comment (rohit7195) [reply java/flight/pom.xml]: So it seems like we import specific "sub-packages" and when I bumped the version up, it required this package. I think the cleaner way to do it would to use grpc-all (https://mvnrepository.com/artifact/io.grpc/grpc-all)

Comment (jacques-n) [reply java/flight/pom.xml]: Let's avoid doing all unless we need to

File: java/flight/pom.xml (hunk 8 above)
Comment (jacques-n) [reply java/flight/pom.xml]: We don't really have one for experimental things. Do you think we should avoid the bump?

**Review by jacques-n — COMMENTED** [review]
Changes look good. I'm +1 once others @lidavidm and @emkornfield are happy with the patch.

File: java/flight/pom.xml (hunk 6 above)
Comment (rohit7195) [reply java/flight/pom.xml]: This is the error while compiling if we remove the dependency:

[WARNING] Used undeclared dependencies found:
[WARNING]    io.grpc:grpc-api:jar:1.24.0:compile
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  35.102 s
[INFO] Finished at: 2019-10-08T12:46:08-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:3.0.1:analyze-only (analyze) on project arrow-flight: Dependency problems found -> [Help 1]

File: java/flight-grpc/src/test/java/org/apache/arrow/flight/TestFlightGrpcUtils.java (hunk 7 above)
Comment (rohit7195) [reply java/flight-grpc/src/test/java/org/apache/arrow/flight/TestFlightGrpcUtils.java]: Refactored to assertThrow() with FlightRuntimeException

File: java/flight/pom.xml (hunk 6 above)
Comment (jacques-n) [reply java/flight/pom.xml]: api is the right import, let's keep this as was done. Someone added a dependency check step that ensures: If we refer to something specifically, we need to include it as a dependency, even if it is already included as a transitive dependency (to answer @emkornfield's question).

File: java/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+  /**
+   * Creates a Flight service.
+   * @param allocator Memory allocator
+   * @param producer Specifies the service api
+   * @param authHandler Authentication handler
+   * @param executor Executor service
+   * @return FlightBindingService
+   */
+  public static FlightBindingService createFlightService(BufferAllocator allocator, FlightProducer producer,
```
Comment (jacques-n) [inline java/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java]: I think should return BindableService. Any reason to expose FlightBindingService specifically?

Comment (rohit7195) [reply java/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java]: Done

File: java/flight/pom.xml (hunk 8 above)
Comment (lidavidm) [reply java/flight/pom.xml]: I think this is fine, with the caveat that technically, gRPC 1.24 requires Protobuf 3.7, not the 3.5 version we have pinned. (We're stuck on an older gRPC, but I don't think anything in Flight is particularly version-specific beyond the serialization tricks in ArrowMessage.)

**Comment by lidavidm** [comment]
I'm happy with the patch once Protobuf is also upgraded to the version implied by the gRPC version. (As with gRPC, I don't think anything in Flight is particularly sensitive to the version either way.) I would like to have the larger conversation about how to separate Flight and its backend protocols eventually, but that need not happen here.

**Comment by emkornfield** [comment]
I won't be able to look at this for a couple of days so don't hold up merging if everyone else is happy with it

**Comment by rohit7195** [comment]
@lidavidm updated protobuf version to 3.7.0

**Review by lidavidm — APPROVED** [review]

**Comment by rohit7195** [comment]
protoc 3.7.0 was failing to compile on linux because of dependence on dynamically linked libatomic, which seems to have been resolved in 3.7.1
Ref: https://github.com/protocolbuffers/protobuf/issues/5875

**Comment by rohit7195** [comment]
@lidavidm @jacques-n @emkornfield 

Can we override the flaky test & merge? I don't think any of the changes in the patch should be causing it

**Comment by emkornfield** [comment]
I just restarted the flaky test, in looks like there is might be an issue with clang-7 but I agree its probably not related.

**Comment by praveenbingo** [comment]
Thanks @rohit7195 for the patch.

Thanks @emkornfield @lidavidm for the reviews.
